### DISCUSSION TITLE
where is the old aws_smithy_client::{erase::DynConnector, test_connection::TestConnection};

### DISCUSSION THREAD

**Discussion by ymwjbxxq** [awslabs/aws-sdk-rust#1038] [Q&A] [answered]
Hello team,

I've been working on updating our code to use a new library, and as part of this process, I've been writing unit tests.

I've successfully updated some parts of the codebase, like changing `aws_smithy_http::body::SdkBody` to `aws_smithy_types::body::SdkBody`. 

However, I've been having trouble finding the equivalent code for `aws_smithy_client::{erase::DynConnector, test_connection::TestConnection}`

 I wonder if these components no longer exist or if they've been moved elsewhere. 
 
Could you please help me locate them?
 
```
aws-smithy-runtime = "1.1.2"
aws-smithy-runtime-api = "1.1.2"
```

The old code:

```
use aws_config;
use aws_sdk_cloudfront;
use aws_sdk_dynamodb;
use aws_sdk_sqs;
use aws_smithy_client::{erase::DynConnector, test_connection::TestConnection};
use aws_smithy_http::body::SdkBody;
use http::Request;
use lambda_http::Response;

pub struct UnitTestHelper {}

impl UnitTestHelper {
    pub fn buil_test_connnection(
        request: Option<Request<SdkBody>>,
        response: Option<Response<SdkBody>>,
    ) -> TestConnection<SdkBody> {
        let mut event = Self::default_event();
        if let (Some(request), Some(response)) = (request, response) {
            event.0 = request;
            event.1 = response;
        }
        TestConnection::new(vec![(event.0, event.1)])
    }
    pub async fn dynamo_fake_client(conn: &TestConnection<SdkBody>) -> aws_sdk_dynamodb::Client {
        let conf = aws_sdk_dynamodb::Config::builder()
            .http_connector(DynConnector::new(conn.clone()))
            .credentials_provider(aws_sdk_dynamodb::config::Credentials::new(
                "accesskey",
                "privatekey",
                None,
                None,
                "dummy",
            ))
            .region(aws_sdk_dynamodb::config::Region::new("eu-central-1"))
            .build();

        aws_sdk_dynamodb::Client::from_conf(conf)
    }
 ```

With this helper class, I do unit tests like this.
```
//something
 let conn = UnitTestHelper::buil_test_connnection(Some(request), Some(response));
let dynamo_db_client = UnitTestHelper::dynamo_fake_client(&conn).await;
//something
    
```

Unless, of course, there is a better way to do all this.

Thanks for any help

**Comment by rcoh** [comment, marked as answer]
Ah good question. These were renamed to be a little bit more specific to their functionality: https://docs.rs/aws-smithy-runtime/latest/aws_smithy_runtime/client/http/test_util/struct.StaticReplayClient.html

Dyn connector shouldn't be required anymore. You can pass `StaticReplayClient` directly as the argument to `http_client`. Here's an [example](https://github.com/smithy-lang/smithy-rs/blob/ed8763e5df356d20b8b8df116c5e4a033f9b1b22/aws/sdk/integration-tests/ec2/tests/paginators.rs#L31) from our own tests

**Reply by ymwjbxxq** [reply]
Why are you using http = "0.2.0" and not 1?
With 0.2 it is all working

with 1.0 instead

The `ReplayEvent`  takes a request and a response of type: `http::Request` and `http::Response`
but even the code of the example:
`http::Request::builder().uri("http://localhost:1234/foo").body(SdkBody::empty()).unwrap()`
`http::Response::builder().status(200).body(SdkBody::empty()).unwrap()`

Gives an error:

```
the trait bound `aws_smithy_runtime_api::http::Request: std::convert::From<http::Request<aws_smithy_types::body::SdkBody>>` is not satisfied
required for `http::Request<aws_smithy_types::body::SdkBody>` to implement `std::convert::Into<aws_smithy_runtime_api::http::Request>`
required for `aws_smithy_runtime_api::http::Request` to implement `std::convert::TryFrom<http::Request<aws_smithy_types::body::SdkBody>>`
required for `http::Request<aws_smithy_types::body::SdkBody>` to implement `std::convert::TryInto<aws_smithy_runtime_api::http::Request>`
```

Would you happen to know why and how to solve it?

**Reply by rcoh** [reply]
http 1.0 came out very shortly before our GA release. We'll be updating the code to support it in the coming weeks.

**Reply by rcoh** [reply]
tracking: https://github.com/smithy-lang/smithy-rs/issues/3365

For some more context, in preparation for this issue, the SDK actually uses our own representation of Requests and Responses in SDK code rather than `http` directly. This allows us to support both versions in a backwards compatible way.
